### PR TITLE
Transfer project ownership

### LINKED ISSUE #1560: feat: Ability to transfer project ownership
**Is your feature request related to a problem? Please describe.**
The project owner is the user who first creates the LF project.  Currently, it is not possible to transfer or change ownership to another user.  The only difference between a project owner and a project manager is:

- Owners cannot remove themselves from the project.  Managers can.
- Owners are the only user who can delete the project.  Even site admins cannot delete the project.

Not being able to transfer project ownership creates a problem for tech support roles who just want to fix a user's problem by deleting the project and starting over from LD.  #1561  addresses allowing admins to delete a project even if they are not a project owner.

**Describe the solution you'd like**
Project ownership is currently reflected in the Project Settings page and in the Share / User Access modal dialog.

- [x] Remove reference to product owner in the Project settings tab.  This was put there initially before the share dialog displayed the owner.
- [x] In the roles dropdown next to each member, add a "Make Owner" link at the bottom of the list, above remove.
- [x] Display a confirm dialog (is this possible since we're already in a modal?) for transferring ownership.

**Additional thoughts for the implementer**

- [x] The "make owner" option in the roles dropdown is only present if the current user is the owner OR if the current user is a site admin.  Otherwise, this option is not visible
- [x] Let's have an icon for the "make owner" action.  Maybe an up arrow?
- [x] Let's add a red X icon for remove, the action below
- [x] Let's add a confirm dialog for the remove user action as well - currently there is not confirmation to remove someone - it just happens instantly.
- [x] I don't know if we should have a new transfer_ownership API endpoint to facilitate this change, or just build it into the change role endpoint.  I'm open to suggestions.
- [x] Only the current owner or a site admin can choose a new owner.  If the current user transfers ownership to another user, the current user becomes a manager
- [x] The application would need to be refreshed after changing ownership to reflect the new permissions in the app

**Describe alternatives you've considered**
We could provide a "transfer ownership" dropdown in the project settings where the "project owner" label is currently being displayed.  I decided that it's probably better to keep user roles UI all together in the Share Dialog.

**Mockups**

New Project settings page without owner info:
Screen Shot 2022-10-26 at 11 23 35 AM

Share Dialog with "make owner" in roles dropdown
Screen Shot 2022-10-26 at 11 24 49 AM

### LINKED ISSUE #1561: feat: allow admins to delete a project
**Is your feature request related to a problem? Please describe.**
Currently, only project owners can delete a project.  This is challenging from a tech support perspective.

**Describe the solution you'd like**
If you are a site admin, you should be able to delete a project.

**Describe alternatives you've considered**
This feature request is related to #1560 

**Additional context**
This feature will be helpful when helping users to "reset" their project from Language Depot

image
The Delete tab should work for site admins, not just project owners

### PR DESCRIPTION
**Pull request by laineyhm**
Fixes #1560 
Fixes #1561 

## Description

Project owners and site admins will now have an option to transfer ownership of a project to another member of the project. We added tests for this to the unit tests.

We also added a red X icon to the "remove" user option, and added confirm dialogs to both the "make owner" and "remove user" actions.

### Type of Change

Only keep lines below that describe this change, then delete the rest.

- New feature (non-breaking change which adds functionality)

## Screenshots

View of an admin:
image

After making someone else owner:
image

View of an owner (non-admin):
image

Confirm dialog ("chicken page" 🐔 ):
image

Transferred ownership back to "admin"; someone who is not an owner or admin will not see the option to transfer ownership:
image

Cannot transfer to an invitee:
image

Confirm dialog for removing a user:
image

Here you can see that this admin is not the owner of the project:
image

but the admin is still shown the delete tab:
image

which the admin uses and then the project is deleted:
image

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have reviewed the title/description of this PR which will be used as the squashed PR commit message
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works

## How to test

I. As an owner, share to another user. Verify that until the other user signs in, you cannnot make them a project owner. Make the other user a manager. Sign in as the other user. This user (a manager) should not be given the option to transfer project ownership. This user should not be shown the delete tab in project settings.
2. Now, as the owner again, make the other user an owner. Sign in as the other user. Verify that this new owner can transfer ownership as well. 
3. Now sign in as an admin and make/view a project. Verify that you can see the project delete tab. Then share to another user. Verify that you can't transfer project ownership before the other user signs in an joins the project. After the user signs in and joins the project, verify that you (an admin) can transfer project ownership to them.

## qa.languageforge.org testing

Testers should add his/her findings to end of the PR in a comment and include screenshots, files, etc that are beneficial.

### CODE REVIEW

**Comment by laineyhm** [comment]
** I realized that I haven't given the new owner all the permissions they need. Even though their id becomes the owner ref in the project, they don't automatically get all the owner rights. I'll message when it's ready for review! **

File: src/Api/Model/Shared/Command/ProjectCommands.php
Comment (longrunningprocess) [inline src/Api/Model/Shared/Command/ProjectCommands.php:257]: really well written 🏆

File: src/Api/Model/Shared/Command/ProjectCommands.php
Comment (longrunningprocess) [inline src/Api/Model/Shared/Command/ProjectCommands.php:275]:
```suggestion
        // get the id of the previous owner
```

File: src/Api/Model/Shared/ProjectModel.php
Comment (longrunningprocess) [inline src/Api/Model/Shared/ProjectModel.php:193]: you can remove this.

File: src/angular-app/bellows/core/api/json-rpc.service.ts
Comment (longrunningprocess) [inline src/angular-app/bellows/core/api/json-rpc.service.ts:82]: let's get rid of this

**Review by longrunningprocess — CHANGES_REQUESTED** [review]
nice work, a few minor things before approval.

File: src/Api/Model/Shared/ProjectModel.php
Comment (longrunningprocess) [reply src/Api/Model/Shared/ProjectModel.php:193]: I only meant to get rid of the comment line, sorry about that.

Comment (laineyhm) [reply src/Api/Model/Shared/ProjectModel.php:193]: So, in this PR, both lines were taken out. Chris said that the instance() call isn't doing anything. From my testing it seems like it's OK to remove it

File: src/angular-app/bellows/core/api/json-rpc.service.ts
Comment (laineyhm) [reply src/angular-app/bellows/core/api/json-rpc.service.ts]: thumbs up

File: src/Api/Model/Shared/ProjectModel.php
Comment (longrunningprocess) [reply src/Api/Model/Shared/ProjectModel.php:193]: yeah, sorry for the confusion, I shouldn't said anything, just wasn't paying close attention 🤪

**Comment by longrunningprocess** [comment]
Lainey noticed she was missing some code on this, so I'm moving it back to draft.

File: src/angular-app/bellows/core/api/json-rpc.service.ts
Comment (myieye) [inline src/angular-app/bellows/core/api/json-rpc.service.ts]: This looks forgotten.

File: src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts
Comment (myieye) [inline src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts:71]: Since `this.session.hasSiteRight(this.sessionService.domain.USERS, this.sessionService.operation.VIEW)` is a bit cryptic, how about adding a new function to `session.service.ts` for this:
```
isSystemAdmin() {
  return this.hasSiteRight(this.sessionService.domain.USERS, this.sessionService.operation.VIEW);
}
```

File: test/php/model/shared/commands/ProjectCommandsTest.php
Comment (myieye) [inline test/php/model/shared/commands/ProjectCommandsTest.php:753]: I like your tests 👍

File: src/Api/Model/Shared/Rights/SiteRoles.php
Comment (myieye) [inline src/Api/Model/Shared/Rights/SiteRoles.php:11]: `project_creator` is a string stored in Mongo. I would guess that changing it here will break things.
Renaming the variable sounds like a good idea though. 👍

image

**Review by myieye — CHANGES_REQUESTED** [review]

File: src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts
Comment (laineyhm) [reply src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts:77]: I like this suggestion, and think we could make role-checking easier for other roles as well through methods like this. But what do you think @megahirt @longrunningprocess ? The "hasSiteRight" method with the domain and operation params seems to be the way role-checking occurs throughout the project right now

File: src/Api/Model/Shared/Rights/SiteRoles.php
Comment (laineyhm) [reply src/Api/Model/Shared/Rights/SiteRoles.php]: Thanks, great point. I changed it back. For me, though, I'm still seeing some potential issues with the permissions in the case that you are showing (an empty project). Working on that...

Comment (laineyhm) [reply src/Api/Model/Shared/Rights/SiteRoles.php]: Okay, from testing on production (languageforge.org), I see that we have some quirks in how the editor component appears when an empty project is shared with an editor or commentor (but not a manager); the quirks that I'm seeing are happening on production and aren't a result of the PR, so I'll make a separate issue.

File: src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts
Comment (myieye) [reply src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts:77]: How did you know that `hasSiteRight(USERS, VIEW)` was the right way to check this?

**Review by myieye — APPROVED** [review]

File: src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts
Comment (laineyhm) [reply src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts:77]: I'm finding issues on my other PR with permissions checking so I will revisit this as well

Comment (megahirt) [reply src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts]: I won't go into the history of why we have permissions distinct from roles (the need for that in our current system is debatable and probably unnecessary complexity) but essentially the permission sets (domain + operation) are hard-coded in an [initialization function in PHP called SystemRoles.php](https://github.com/sillsdev/web-languageforge/blob/chore/removeWebsiteClass/src/Api/Model/Shared/Rights/SystemRoles.php#L27).  If you understand this function (which is regrettably somewhat hard to read), then you can know that admins have a site right of User+View.

File: test/php/model/shared/commands/ProjectCommandsTest.php
Comment (megahirt) [reply test/php/model/shared/commands/ProjectCommandsTest.php:753]: Yes!!!  Great tests!

File: src/angular-app/languageforge/lexicon/shared/share-with-others/role-dropdown.component.html
Comment (megahirt) [inline src/angular-app/languageforge/lexicon/shared/share-with-others/role-dropdown.component.html:11]: > `ng-if="$ctrl.currentUserIsOwnerOrAdmin() && !userIsInvitee(`
To me this line tells me that the "make owner" option is not visible if the user is only "invited".

I'm confused then how one would get to the "chicken page" where it says invitees are not allowed to be owners (yet).

File: src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts
Comment (megahirt) [inline src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts:76]: If I'm reading the conditional logic right, we should never hit this code since we don't show the `make owner` option for invitees.  Can you explain how we would get here?

And even then, I don't like having a modal that pops up to tell the user they cannot do something.  We should just not have the option all together in that case.

**Review by megahirt — CHANGES_REQUESTED** [review]
Well done @laineyhm! This is certainly your biggest contribution yet and one that crosses into many aspects of our application, including PHP tests!

I have a few things that need changing before approval, notably:
- invitees chicken page - users shouldn't get here since they shouldn't be able to make an invitee the owner
- fix up the extra permission you added
- refactor a bit of code

File: src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts
Comment (megahirt) [reply src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts:76]: I suggest deleting this block.

**Comment by megahirt** [comment]
@laineyhm you will want to copy over the screenshots and testing instructions from #1607 to this PR, since this is now "the PR to rule them all" :)

File: src/Api/Model/Shared/Command/ProjectCommands.php
Comment (laineyhm) [reply src/Api/Model/Shared/Command/ProjectCommands.php:278]: The currentUser might be an admin, in which case the previous owner is a different user. The previous owner always needs their role updated, even if they are not doing the transfer, but rather an admin.

Comment (megahirt) [reply src/Api/Model/Shared/Command/ProjectCommands.php:278]: ha!  You're right.  Then we do need the code as written in this case.  Thanks for explaining that.  Disregard the following changes as well then.

File: src/Api/Model/Shared/Command/ProjectCommands.php
Comment (megahirt) [reply src/Api/Model/Shared/Command/ProjectCommands.php:285]: you're right!  I had gotten the wrong idea in my head about certainty that currentUser and previousOwner were always the same.  So this suggestion isn't needed.

File: src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts
Comment (megahirt) [reply src/angular-app/languageforge/lexicon/shared/share-with-others/user-management.component.ts]: deleted in latest commit

File: src/angular-app/languageforge/lexicon/shared/share-with-others/role-dropdown.component.html
Comment (megahirt) [reply src/angular-app/languageforge/lexicon/shared/share-with-others/role-dropdown.component.html]: this was removed in the latest commit

File: src/Api/Model/Shared/Dto/RightsHelper.php
Comment (laineyhm) [reply src/Api/Model/Shared/Dto/RightsHelper.php]: Got it, made this change

File: src/Api/Model/Shared/Rights/SiteRoles.php
Comment (laineyhm) [reply src/Api/Model/Shared/Rights/SiteRoles.php]: That clears things up, thanks. Made the change

File: src/Api/Model/Shared/Rights/SiteRoles.php
Comment (laineyhm) [reply src/Api/Model/Shared/Rights/SiteRoles.php]: 👍

File: src/Api/Model/Shared/Rights/SiteRoles.php
Comment (laineyhm) [reply src/Api/Model/Shared/Rights/SiteRoles.php]: 👍

File: src/Api/Model/Shared/Rights/SiteRoles.php
Comment (megahirt) [reply src/Api/Model/Shared/Rights/SiteRoles.php]: done in latest commit

**Comment by longrunningprocess** [comment]
# Testing in staging

💡  don't forget about the gmail `+` hack for testing, e.g., <your-email>+test-1@gmail.com

## un-verified user cannot be made project owner

**Comment by laineyhm** [comment]
@longrunningprocess I adjusted your list of test cases slightly, and I think it is now complete. @megahirt

**Comment by laineyhm** [comment]
This finishes the tests, I think.  I have also tested for the cases above with the same results. 🎉 

## verify admin user can transfer ownership to another user after they have been verified ✅  ## 
image
image

## verify admin user can see the delete tab in project settings ✅ ##
image

**Comment by longrunningprocess** [comment]
excellent, thank you 🤩

**Pull request merged**
